006-restarting-kafka-connect-connectors-and-tasks.md

```diff
@@ -0,0 +1,116 @@
+
+
+## Motivation
+
+Enabling users to restart connectors and tasks means there are almost (1) no further Connect REST APIs that users must call to interact with connectors/tasks when using CRs that create connectors. This means the CRs and their operators can exclusively be used to manage connectors and tasks - Strimzi users do not have to call the Connect REST API directly. As a corollary, this means the Connect REST listener does not need to be available to users; it could be secured so that only the operator can interact with the REST API.
+
+(1) - Kafka 2.5 introduced the `GET /connectors/{connector-name}/topics` Connect endpoint to list the topics that a connector has used since the connector was created or the list was reset. Kafka 2.5 also introduced the `PUT /connectors/{connector-name}/topics/reset` Connect endpoint to clear the list of topics that a connector has used.
+
+
+## Proposal
+
+This proposal adds two new separate but related features:
+
+1. Two new annotations that cause the operator to restart connectors or tasks. They can be applied to the `KafkaConnector`, and the `KafkaMirrorMaker2` custom resources. The annotation acts as a trigger for a single restart call by the operator, and is removed from the CR when the restart REST API call is successfully called.
+
+2. New default behaviour in the `KafkaConnector` and `KafkaMirrorMaker2` operators that automatically attempts to restart connectors and tasks if they are in the `FAILED` state. The `KafkaConnector` and `KafkaMirrorMaker2` CRs can use a new annotation to disable the automatic restarting behaviour if required.
+
+These two features could be delivered separately as they do not rely on each other.
+
+
+### `strimzi.io/restart` annotation
+
+An annotation named `strimzi.io/restart` can be applied to the `KafkaConnector` and `KafkaMirrorMaker2` custom resources. The value of the annotation and the CR it is applied to determines the connector that is restarted. Once the operator successfully calls the restart REST API endpoint, the CR is patched to remove the annotation. The operator attempts to call the restart REST API endpoint once per reconcile cycle.
+
+When applied to a `KafkaConnector` CR, the connector defined by the CR is restarted. Any value can be assigned to the annotation. For example:
+```
+apiVersion: eventstreams.ibm.com/v1alpha1
+kind: KafkaConnector
+metadata:
+  name: my-source-connector
+  annotations:
+    strimzi.io/restart: "true"
+  ...
+```
+
+When applied to a `KafkaMirrorMaker2` CR, the annotation value contains the name of the connector that is restarted. For example:
+```
+apiVersion: eventstreams.ibm.com/v1alpha1
+kind: KafkaMirrorMaker2
+metadata:
+  name: my-mm2-cluster
+  annotations:
+    strimzi.io/restart: "source-cluster->target-cluster.MirrorSourceConnector"
```

Comment:
so ... while `strimzi.io/restart` makes a lot of sense to me on a `KafkaConnector` resource because it's clear that you are asking to restart the "resource" so the connector, it's not so clear on a `KafkaMirrorMaker2` resource where it seems that your intention is to restart MM2. Maybe we can change this annotation in `strimzi.io/restart-connector` ?

Reply:
👍 I have added `strimzi.io/restart-connector` for `KafkaMirrorMaker2` CRs